stresstests/test/stress_measurements_recorder.js

```diff
@@ -0,0 +1,143 @@
+  }
+
+  /**
+   * This should be called when a message is sent from the device side to keep
+   * track of the number of messages in the queue.
+   * @public
+   */
+  messageEnqueued() {
+    if (this._done) {
+      throw new Error('start() must be called before calling messageArrived()');
+    }
+    this.peakTelemetryMessagesInQueue = Math.max(this.peakTelemetryMessagesInQueue, ++this.telemetryMessagesInQueue);
+  }
+
+  /**
+   * This should be called when a message is received from the service side to
+   * keep track of the messages in the queue and the peak telemetry arrival time.
+   * @param {number} arrivalTimeInMs - The amount of time from when the message
+   *   was sent on the device side to when it was seen on the service side.
+   * @public
+   */
+  messageArrived(arrivalTimeInMs) {
+    if (this._done) {
+      throw new Error('start() must be called before calling messageArrived()');
+    }
+    if (!arrivalTimeInMs) {
+      throw new Error('arrivalTimeInMs must be defined.');
+    }
+    if (typeof arrivalTimeInMs !== 'number') {
+      throw new TypeError('arrivalTimeInMs must be a number');
+    }
+    --this.telemetryMessagesInQueue;
+    this.peakTelemetryArrivalTimeInMs = Math.max(this.peakTelemetryArrivalTimeInMs, arrivalTimeInMs);
+  }
+
+  /**
+   * This function should be called at the beginning of every test.
+   * @param {object} client - A Client or ModuleClient instance. Used to keep
+   *   track of connects and disconnects
+   * @param {number} memorySamplingIntervalInMs - The time interval to sample
+   *   the RSS to keep track of memory usage.
+   * @public
+   */
+  start(client, memorySamplingIntervalInMs) {
+    if (!this._done) {
+      throw new Error('Already started. done() must be called before calling this function.')
+    }
+    if (!memorySamplingIntervalInMs) {
+      throw new Error('memorySamplingIntervalInMs must be defined.');
+    }
+    if (typeof memorySamplingIntervalInMs !== 'number') {
+      throw new TypeError('memorySamplingIntervalInMs must be a number');
+    }
+    if (!client) {
+      throw new Error('client must be defined.');
+    }
+    if (typeof client !== 'object') {
+      throw new TypeError('client must be an object');
+    }
+    this._client = client;
+    this._done = false;
+    this.peakRssInBytes = process.memoryUsage().rss;
+    this.telemetryMessagesInQueue = 0;
+    this.peakTelemetryMessagesInQueue = 0;
+    this.peakTelemetryArrivalTimeInMs = null;
+    this.peakReconnectTimeInMs = null;
+    this.totalElapsedTime = null;
+    this._startTime = Date.now();
+    this._disconnectedTime = null;
+    this._memorySamplingInterval = setInterval(() => {
+      const rss = process.memoryUsage().rss;
+      this.peakRssInBytes = Math.max(this.peakRssInBytes, rss);
+      debug(`RSS: ${rss / 1024 / 1024} MiB (max ${this.peakRssInBytes / 1024 / 1024} MiB)`);
+    }, memorySamplingIntervalInMs);
+    this._client.on('connect', this._connectHandler);
```

Comment:
>_connectHandler [](http://example.com/codeflow?start=36&length=15)

dumb question: why do you have to bind when assigning to `Deferred.then` but not here?

Reply:
Bind doesn't work for arrow functions. In arrow functions, `this` gets captured from the scope that the arrow function is defined in. Though the way I've done this is wrong, because `this` in `_connectHandler` and `_disconnectHandler` will be the global object.